### DISCUSSION TITLE
Support npm registries with a general hostRule authentication

### DISCUSSION THREAD

**Discussion by Chumper** [renovatebot/renovate#24892] [Suggest an Idea]
**What would you like Renovate to be able to do?**

TLDR: I want Renovate to add npm registry authentication information to the npmrc config of a repository.

This behaviour has changed with `v25` of Renovate, when the decision was made to deprecate the `NPM_TOKEN` support.
Let's assume the following `npmrc` in the `renovate.json` file:
```
registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-release/
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:always-auth=true
@company:registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-public-release
```

and the following `hostRules` entry in the bot config:
```
        {
            "baseUrl": "artifactory.corp.company.com",
            "hostType": "npm",
            "password": "ARTIFACTORY_TOKEN",
            "username": "renovate"
        },
```

When Renovate works on the repository, it will add the following authentication to the in-memory npmrc:
```
registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-release/
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:always-auth=true
@company:registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-public-release

//artifactory.corp.company.com/:_authToken=<token>
```

The problem is that in this case it is not enough to add authentication for the host itself.
The authentication needs to be scoped to the registries.
In the end I would expect Renovate to create the following in-memory `npmrc`:
```
registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-release/
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:always-auth=true
@company:registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-public-release

//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:_authToken=<token>
```

That would unblock us from switching to `v25`.

For Renovate `<v25` the following was working without any problems:
```
registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-release/
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:_authToken=${NPM_TOKEN}
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:email=${NPM_EMAIL}
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:always-auth=true
# Registry for custom package
@company:registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-public-release
```

Here we relied on the `NPM_TOKEN` replacement to work, that is not possible anymore.

**Did you already have any implementation ideas?**

Renovate should not only add the `hostRule` to the npmrc but instead add all `registriesUrls` with corresponding auth where any of the `hostRules` matches.
This would solve the problem that all registries would have to be added by hand.
Instead a hostRule for the domain can be created and Renovate can support all npm registries on that domain.

As the registryUrl is already known for the dependencies Renovate simply needs to create a map of registries + authentication and add them to the in-memory `npmrc`

I think the correct place to iterate over the registries is here: https://github.com/renovatebot/renovate/blob/0918679cf6d12f0b96799e829afa047924e9d376/lib/manager/npm/post-update/index.ts#L441

This is already the case with docker:
```
        {
            "domainName": "corp.company.com",
            "hostType": "docker",
            "password": "ARTIFACTORY_CORP",
            "username": "renovate"
        },
```
The above `hostRule` supports all docker registries under the domain name, like `my-reg.corp.company.com` o.s. 
We want the same for npm.

**Comment by viceice** [comment]
Ref https://github.com/renovatebot/renovate/issues/9145

**Comment by viceice** [comment]
https://github.com/renovatebot/renovate/issues/5646#issuecomment-595071930

**Comment by viceice** [comment]
Strange, i thought it wqould be enough for `yarn` to have the folling `.npmrc` to sucessfully authenticate 

```ini
//artifactory.corp.company.com/:_authToken=XXXX
registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-public-release
always-auth=true
```
---
So that needs to configure this in renovate:
hostRule:
```js
       {
            "baseUrl": "artifactory.corp.company.com",
            "hostType": "npm",
            "token": "<ARTIFACTORY_TOKEN>",
        },
```

and the following in repo `.npmrc`:
```ini
registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-public-release
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:_authToken=${NPM_TOKEN}
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:email=${NPM_EMAIL}
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:always-auth=true
```

Repo `.npmrc` can have lines with env tokens, they will silently dropped by renovate before using.

This config should create the folling npmrc by renovate:
```ini
registry=https://artifactory.corp.company.com/artifactory/api/npm/npm-company-public-release
//artifactory.corp.company.com/artifactory/api/npm/npm-company-release/:always-auth=true
//artifactory.corp.company.com/:_authToken=<ARTIFACTORY_TOKEN>
```

**Comment by jamie-tm** [comment]
My `config.js` file has something like this:
```js
const artifactoryBaseUrl = 'artifactory.mycompany.com';
const { artToken, artUsername, artPassword } = externalCredentials;
module.exports = {
    hostRules: [
        {
            // Authenticate Artifactory NPM repos.
            hostType: 'npm',
            matchHost: `${artifactoryBaseUrl}/artifactory/api/npm`,
            token: artToken,
            authType: 'Basic'
        },
        {
            // Authenticate other types of Artifactory repos as well.
            matchHost: artifactoryBaseUrl,
            username: artUsername,
            password: artPassword
        }
    ],

    npmrc: [
            `registry=https://${artifactoryBaseUrl}/npm-remote`,
            `@private:registry=https://${artifactoryBaseUrl}/private-npm-local/`
    ].join('\n')
}
```

At run-time, this results in a temporary `.npmrc` file that looks something like this:
```
registry=https://artifactory.mycompany.com/artifactory/api/npm/npm-remote
@private:registry=https://artifactory.mycompany.com/artifactory/api/npm/private-npm-local/

//artifactory.mycompany.com/artifactory/api/npm/:_auth=INJECTED_AUTH_TOKEN
```

Note that the injected `_auth` configuration is for a _prefix_ of the specified registries; it doesn't include the actual repository name at the end. (And, actually, it could be `//artifactory.mycompany.com/` instead just as easily and still work the same way.)

I've found that this works with the latest version of NPM, but fails if I use an older one. I managed to narrow it down. The earliest working version I've found is npm `7.11.0`. `7.10.2` fails to authenticate. Looking at the [change notes](https://github.com/npm/cli/releases/tag/v7.11.0), it looks like it might be related to the change to `npm-registry-fetch` version. I think it's [this change](https://github.com/npm/npm-registry-fetch/commit/cc11cc15c9f457c01159a69ce342dc99dea12456) that does it.

So, good news is that URL prefixes (such as `//artifactory.corp.company.com/:_authToken=<ARTIFACTORY_TOKEN>`) _do_ work, but only for newer versions of NPM.

To support older versions of NPM, it would be nice if Renovate looked for matching configured registries and provide the token to each of them -- as describe in the original issue.

**Comment by markusheiden** [comment]
Please consider supporting scoped registries for yarn/.yarnrc too.
